### PR TITLE
SAS7BDAT parser: Fast byteswap

### PR DESCRIPTION
**Pull request by jonashaag**
Speed up SAS7BDAT int/float reading.

This is order of magnitude faster than using `struct.unpack(fmt, data)` or `precompiled_unpacker = struct.Struct(fmt).unpack; ...; precompiled_unpacker(data)`.

Unfortunately Python does not expose a low-level interface to `struct` or a byteswapping interface. The byteswap implementation in this change is from `pyreadstat`.

Today this brings a modest 10-20% performance improvement. But together with the other changes I will be proposing it will be a major bottleneck.

- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/sas/sas.pyx [hunk 1]
```diff
@@ -1,13 +1,19 @@
 # cython: profile=False
 # cython: boundscheck=False, initializedcheck=False
 from cython cimport Py_ssize_t
+from libc.stdint cimport (
+    int64_t,
+    uint8_t,
+    uint16_t,
+    uint32_t,
+    uint64_t,
```
Comment (jbrockmendel) [inline pandas/io/sas/sas.pyx:9]: just curious, are these interchangeable with the versions of these we cimport from numpy?

Comment (jonashaag) [reply pandas/io/sas/sas.pyx:9]: Interesting, I didn't know those exist.

File: pandas/io/sas/sas.pyx [hunk 2]
```diff
@@ -433,3 +439,73 @@ cdef class Parser:
...
+
+cdef inline uint64_t _byteswap8(uint64_t num):
+    num = ((num & <uint64_t>0xFFFFFFFF00000000) >> 32) | ((num & <uint64_t>0x00000000FFFFFFFF) << 32)
+    num = ((num & <uint64_t>0xFFFF0000FFFF0000) >> 16) | ((num & <uint64_t>0x0000FFFF0000FFFF) << 16)
+    return ((num & <uint64_t>0xFF00FF00FF00FF00) >> 8) | ((num & <uint64_t>0x00FF00FF00FF00FF) << 8)
+
+
+cdef inline float _byteswap_float(float num):
+    cdef uint32_t answer = 0
+    memcpy(&answer, &num, 4)
```
Comment (WillAyd) [inline pandas/io/sas/sas.pyx:500]: Can you use sizeof instead of hard coding the size?

Comment (jonashaag) [reply pandas/io/sas/sas.pyx:500]: This is a verbatim copy from ReadStat, do you still want me to make that modification?

Comment (WillAyd) [reply pandas/io/sas/sas.pyx:500]: Ah ok. I think fine to keep as is then

**Comment by jonashaag** [comment]
One thing I realized is that we could also use NumPy's byteswapping, at the cost of around 10% worse performance relative to this impl.

**Comment by jbrockmendel** [comment]
> One thing I realized is that we could also use NumPy's byteswapping, at the cost of around 10% performance relative to this impl.

How much of this could be replaced with that?

**Comment by jonashaag** [comment]
> How much of this could be replaced with that?

Everything copied from read stat so ~ 50 lines of Cython code

File: pandas/io/sas/sas.pyx [hunk 3]
```diff
@@ -433,3 +439,73 @@ cdef class Parser:
         self.current_row_on_page_index += 1
         self.current_row_in_chunk_index += 1
         self.current_row_in_file_index += 1
+
+
+def read_float_with_byteswap(const uint8_t *data, bint byteswap):
```
Comment (jreback) [inline pandas/io/sas/sas.pyx:444]: can you add some comments here on *what* and *why* you are doing this.

**Review by jreback — CHANGES_REQUESTED** [review]
asv's that cover this case?

can you add a whatsnew note

**Comment by jonashaag** [comment]
ASV from #47405:

```
       before           after         ratio
     [e915b0a4]       [435a003c]
     <main>           <sas/byteswap~1>
-      81.9±0.7ms       73.7±0.3ms     0.90  io.sas.SAS.time_read_sas7bdat_2_chunked
-      78.7±0.7ms       69.5±0.6ms     0.88  io.sas.SAS.time_read_sas7bdat_2
```

**Comment by jonashaag** [comment]
Test failure seems unrelated

**Comment by jonashaag** [comment]
I'm trying an intrinsics-based version: Smaller code and generally faster.

**Comment by jonashaag** [comment]
Test failure is unrelated. @jreback merge?

**Comment by datapythonista** [comment]
@jonashaag seems like you didn't add the release note as requested, do you mind adding it?

Does it make sense to add a unit test for the new functions?

Also if you can please merge master.

**Comment by jonashaag** [comment]
@datapythonista release notes are already merged with #47404. Will merge main.

**Comment by jonashaag** [comment]
Do you want me to add unit tests?

**Comment by datapythonista** [comment]
> Do you want me to add unit tests?

Personally I think it should be useful. Probably just one parametrized to test the function for every type is enough.

**Comment by datapythonista** [comment]
The release note in the other PR is not accurate now. The changes in this PR won't be available until pandas 1.6/2.0, while for what you mention the release note in the other PR mentions this issue/PR for 1.5. Probably no big deal, but if you want to open a small PR to update that note in the releases to only include the issues/PRs that have already been merged, that would leave things more accurate (ping me in that PR, so I backport it to 1.5). And then you can add another release note for 1.6 here.

**Comment by jonashaag** [comment]
Or we merge the PRs noted in those release notes to 1.5? The PRs have been ready for review with no code changes for a long time.

**Comment by datapythonista** [comment]
We already released pandas 1.5 release candidate, and we are only backporting regressions to it, not new features, performance improvements. It's also unclear to me this will be merged before the release.

I know this has been forgotten for a long time, sorry about that. But that's unfortunately part of how open source development works in a project like pandas.

**Comment by jonashaag** [comment]
I added tests using Hypothesis and moved the byteswapping code to a module. The byteswapping code also be moved somewhere else outside the SAS stuff.

**Comment by jonashaag** [comment]
Updated release notes. I took the liberty of including #47656 already.

File: pandas/io/sas/sas7bdat.py [hunk 4]
```diff
@@ -349,37 +358,46 @@ def __next__(self) -> DataFrame:
...
+                self._cached_page, offset, self.need_byteswap
+            )
+        else:
             self.close()
             raise ValueError("invalid float width")
-        buf = self._read_bytes(offset, width)
-        fd = "f" if width == 4 else "d"
-        return struct.unpack(self.byte_order + fd, buf)[0]
 
     # Read a single signed integer of the given width (1, 2, 4 or 8).
```
Comment (mroeschke) [inline pandas/io/sas/sas7bdat.py:374]: Is this comment still accurate as your new functions are for unsigned ints? Or should we have e.g. `read_int16_with_byteswap` versions too?

Comment (jonashaag) [reply pandas/io/sas/sas7bdat.py]: Good catch, to the best of our knowledge those numbers are actually unsigned, see https://github.com/Roche/pyreadstat/blob/e0627c7cf406e4296a9af16036a563719fcba237/src/sas/readstat_sas.c#L239-L288

Renamed the function accordingly.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jonashaag

**Pull request merged**